### ISSUE TITLE
La commande RefreshAllMaterializedViews ne fonctionne pas

### ISSUE THREAD

**Issue by georgemoon** [PnX-SI/GeoNature-atlas#254]
La commande : `SELECT RefreshAllMaterializedViews('atlas');` entraîne la réponse suivante :
```
Error occurred during SQL query execution

Motif:
SQL Error [55000]: ERREUR: ne peut pas rafraichir en parallèle la vue matérialisée « atlas.t_layer_territoire »
  Indice : Crée un index unique sans clause WHERE sur une ou plusieurs colonnes de la vue matérialisée.
  Où : instruction SQL « REFRESH MATERIALIZED VIEW CONCURRENTLY atlas.t_layer_territoire »
fonction PL/pgsql refreshallmaterializedviews(text), ligne 10 à EXECUTE
```

(Postgres 9.6.19 sur Debian 9.13 avec une nouvelle installation de GeoNature-Atlas).

Il n'y a aucun problème avec les deux autres commandes (`select atlas.refresh_materialized_view_data();` ou `select atlas.refresh_materialized_view_ref_geo();`).

**Comment by camillemonchicourt** [comment]
Oui en effet, car pour pouvoir rafraîchir une vue matérialisée en parallèle (concurrently) comme réalisé dans la fonction ``RefreshAllMaterializedViews('atlas');``, la vue matérialisée doit contenir un champs basé sur un index unique. A priori ce n'est pas le cas pour ``atlas.t_layer_territoire``. 

Dans ``atlas.refresh_materialized_view_ref_geo()`` la vue matérialisée ``atlas.t_layer_territoire`` n'est pas rafraîchie en parallèle : https://github.com/PnX-SI/GeoNature-atlas/blob/master/data/gn2/atlas_ref_geo.sql#L111.

Désormais, il est conseillé d'utiliser la fonction ``atlas.refresh_materialized_view_data()`` pour le rafraîchissement des données, car elle évite de rafraîchir toutes les vues matérialisées géographiques qui sont longues à rafraîchir et ne sont pas censées bouger régulièrement (territoire, communes...).

Cela fait aussi partie des points transitoires de la version 1.4.

**Comment by georgemoon** [comment]
Ok, merci pour l'explication

**Comment by Avikern** [comment]
Bonjour,

Je me permets de rebondir sur ce post car nous avons un problème avec la même commande.

"ERREUR: valeur trop longue pour le type character varying(250)
État SQL :22001
Contexte : column "lb_auteur" of foreign table "taxref"
instruction SQL « REFRESH MATERIALIZED VIEW atlas.vm_taxref »
fonction PL/pgsql refreshallmaterializedviews(text), ligne 9 à EXECUTE"

Je pense que cela est dû au fait que nous soyons passé à la version 2.4.0 de GeoNature, la nouvelle table _taxonomie.taxref_ étant basée sur la TAXREF 13 et ayant des longueurs de champs différentes de la table _taxonomie.taxref_v11_.

Avez-vous connaissance de problèmes similaires ?

Les deux propositions de commandes faites ci-dessus renvoient un _void_.

Pour info :
GeoNature 2.4.0
GeoNature-Atlas 1.4.1
PostgreSQL 9.6.16
Debian 6.3.0

**Comment by TheoLechemia** [comment]
Bonjour,

En effet, à partir de 2.4.0, il faut Taxhub 1.7.0, qui contient la monté de version vers Taxref 13 : 
Est-ce que vous avez fait la migration vers le v13 ?

**Comment by Avikern** [comment]
Oui, nous avons la version 1.7.2 de TaxHub et la synthèse de GeoNature comprends les recherches taxonomiques du v13 ("_Linaria cannabina_" par exemple, nommée "_Carduelis cannabina_" dans la v11)

Errata : GeoNature est en 2.4.1

**Comment by TheoLechemia** [comment]
Ok, et quel est la longueur du champs "lb_auteur" dans atlas.vm_taxref et taxonomie.taxref ?

**Comment by Avikern** [comment]
L'information n'est pas précisée dans la définition de la vm, ou bien je ne regarde pas au bon endroit (ce qui est possible aussi !).
Mais d'après l'erreur citée précédemment, j'imagine qu'elle est de 250

"lb_auteur character varying(500)" dans taxonomie.taxref

**Comment by TheoLechemia** [comment]
Ok. Il semblerait que le vue materialisé n'est pas répercuté le changement du nombre de caractère de sa table source. 
On pourrait recréer la VM, mais ça nécessiterait de refaire toutes vue dépendantes... une galère.
Une petite astuce qui devrait marcher (pas testé chez moi):

se connecter en ligne de commande avec l'utilisateur postgres:
`sudo su postgres `
se connecter à sa base de l'atlas
`psql -d <NOM_BASE> `

executer cette commande:
```
UPDATE pg_attribute SET atttypmod = 500+4
WHERE attrelid = 'atlas.vm_taxref'::regclass
AND attname = 'lb_auteur';
```

**Comment by TheoLechemia** [comment]
Sinon en solution plus propre il y a ça:
https://github.com/PnX-SI/Ressources-techniques/blob/master/PostgreSQL/vues_dependantes.rst

**Comment by Avikern** [comment]
Ok, merci pour ces pistes et pour votre réactivité.

Je vous tiendrai au courant des résultats

**Comment by Avikern** [comment]
Edit : La première commande ne change pas le résultat de la requête, nous avons toujours la même erreur.

**Comment by TheoLechemia** [comment]
Plus qu'à essayer la 2ème ;)

**Comment by Avikern** [comment]
La deuxième fonctionne parfaitement, voici les champs qui ont nécessité une modification :

- taxref.lb_auteur : 500 caractères -> 250
- taxref.nom_complet : 500 caractères -> 255
- taxref.nom_complet_html : 500 caractères -> 255
- taxref.nom_valide : 500 caractères -> 255

La commande "RefreshAllMaterializedViews" se déroule sans accrocs désormais !

Mais nous n'avons toujours pas données dans atlas.vm_observations...

Je vais continuer de chercher comment remplir notre Atlas, mais merci encore pour le support

**Comment by camillemonchicourt** [comment]
OK donc on a peut-etre des longeurs de champs à ajuster au niveau de GeoNature-atlas.
Concernant la remontée des données dans ``vm_observations``, à voir si ce n'est pas l'intersection avec le territoire qui pose soucis.
C'est une partie qui mériterait d'être simplifiée ou au moins clarifiée dans la documentation. Toute contribution est bienvenue.
En attendant des éléments ont été précisés ici : https://github.com/PnX-SI/GeoNature-atlas/issues/220#issuecomment-565557280

**Comment by JulienCorny** [comment]
Salut,

De mon côté les 2 commandes citées précédemment n'ont pas fonctionné. 
Donc j'ai supprimé atlas.vm_taxref,
Puis :
alter foreign table taxonomie.taxref alter column lb_auteur type character varying(500);
Et là ça fonctionne pour créer atlas.vm_taxref with data

(sans le alter la suppression et re-création de la vue ne va pas fonctionner, la vm reste à character varying(250) lors de la création)

Bon après il faut recréer les vm qui dépendaient de vm_taxref...
